File: sidecar/test/unit/salvage.test.js
'use strict';
// A batch round writes N tests into one file. One bad assertion used to cost all N.
//
// The very first unit of the v2 run proved it: three targets in
// ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest, two of them
// asserting wrongly, the whole file deleted for breaking the suite, and the round recorded
// as cov 0→0 mut 0→0. Under the old one-mutant-per-round loop a bad assertion cost one
// test; batching made it cost the batch.
//
// It does not have to. The file COMPILED — the failures are assertion failures, the
// runner names the methods that failed, and those names are the deterministic target names
// this pipeline chose. So the failing methods can be cut out and the rest kept.
const { test } = require('node:test');
const assert = require('node:assert/strict');
const { failedTestNames, dropTestMethods } = require('../../salvage');
const { stripNonCode } = require('../../javasrc');

// Braces inside string literals are not structure — the fixture below deliberately
// asserts on "}" — so balance is checked over code only. Counting raw characters made
// the UNMODIFIED fixture look unbalanced (4 vs 6) and failed a correct implementation.
const braces = (s) => {
  const c = stripNonCode(s);
  return [(c.match(/\{/g) || []).length, (c.match(/\}/g) || []).length];
};

// captured verbatim from the live v2 run
const GRADLE = `ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest > kill_incrementBatchLoadExceptionCount_line76() FAILED
    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
    at org.dataloader.stats.ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest.kill_incrementBatchLoadExceptionCount_line76(ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest.java:27)
ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest > kill_incrementBatchLoadExceptionCount_line82() FAILED
    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)`;

const SUREFIRE = `[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0
[ERROR] kill_mustEscape_line170(org.json.XMLMacMutTest)  Time elapsed: 0.01 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>`;

test('the failing method names are read out of a Gradle run', () => {
  const f = failedTestNames(GRADLE);
  assert.deepEqual([...f].sort(), [
    'kill_incrementBatchLoadExceptionCount_line76',
    'kill_incrementBatchLoadExceptionCount_line82',
  ]);
});

test('and out of a Surefire run', () => {
  assert.deepEqual([...failedTestNames(SUREFIRE)], ['kill_mustEscape_line170']);
});

test('a green log names nothing', () => {
  assert.equal(failedTestNames('BUILD SUCCESSFUL in 12s').size, 0);
  assert.equal(failedTestNames('').size, 0);
  assert.equal(failedTestNames(null).size, 0);
});

// ── cutting the bad methods out ───────────────────────────────────────────
const FILE = `package a;

import org.junit.jupiter.api.Test;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.equalTo;

class BMacMutTest {

    @Test
    void kill_m_line10() {
        assertThat(new B().m(), equalTo(1));
    }

    @Test
    void kill_m_line20() {
        // this one asserts the wrong thing
        assertThat(new B().m(), equalTo(99));
    }

    @Test
    void kill_m_line30() {
        assertThat(new B().n("}"), equalTo("}"));
    }
}
`;

test('only the named method is removed', () => {
  const out = dropTestMethods(FILE, new Set(['kill_m_line20']));
  assert.doesNotMatch(out, /kill_m_line20/);
  assert.match(out, /kill_m_line10/);
  assert.match(out, /kill_m_line30/);
});

test('the file still has its package, imports and class', () => {
  const out = dropTestMethods(FILE, new Set(['kill_m_line20']));
  assert.match(out, /^package a;/);
  assert.match(out, /import org\.junit\.jupiter\.api\.Test;/);
  assert.match(out, /class BMacMutTest \{/);
  const [o, c] = braces(out);
  assert.equal(o, c, 'braces balance');
});

test('a brace inside a string literal does not end the method early', () => {
  // kill_m_line30 asserts on "}" — counting braces naively would cut the file in half
  const out = dropTestMethods(FILE, new Set(['kill_m_line30']));
  assert.doesNotMatch(out, /kill_m_line30/);
  assert.match(out, /kill_m_line10/);
  assert.match(out, /kill_m_line20/);
  const [o, c] = braces(out);
  assert.equal(o, c, 'braces balance');
});

test('removing every method leaves a compilable empty class, not a broken file', () => {
  const out = dropTestMethods(FILE, new Set(['kill_m_line10', 'kill_m_line20', 'kill_m_line30']));
  assert.match(out, /class BMacMutTest \{/);
  assert.doesNotMatch(out, /@Test/);
  const [o, c] = braces(out);
  assert.equal(o, c);
});

test('naming a method that is not there changes nothing', () => {
  assert.equal(dropTestMethods(FILE, new Set(['kill_m_line999'])), FILE);
  assert.equal(dropTestMethods(FILE, new Set()), FILE);
});

// ── one decision, used by both places that drop tests ─────────────────────
// delete-many learned to cut only the failing methods. The PIT green-suite recovery did
// not, and deleted the whole file — so a batch that survived a red suite by salvage was
// then wiped whole when PIT refused to run against one of its members. Two callers, one
// decision, or they drift.
const { salvageSource } = require('../../salvage');

test('cutting the named methods returns the file when something still runs', () => {
  const out = salvageSource(FILE, new Set(['kill_m_line20']));
  assert.ok(out);
  assert.doesNotMatch(out, /kill_m_line20/);
  assert.match(out, /kill_m_line10/);
});

test('cutting everything returns null — there is nothing worth keeping', () => {
  // a file of one test whose only test failed is a file to delete, not to keep empty
  const out = salvageSource(FILE, new Set(['kill_m_line10', 'kill_m_line20', 'kill_m_line30']));
  assert.equal(out, null);
});

test('naming nothing that is in the file returns null, not a no-op keep', () => {
  // the caller asked us to drop something we cannot find — deleting is the safe answer,
  // since the file demonstrably broke the build and we cannot say which part
  assert.equal(salvageSource(FILE, new Set(['kill_other_line1'])), null);
  assert.equal(salvageSource(FILE, new Set()), null);
});

test('a file we cannot read is not salvageable', () => {
  assert.equal(salvageSource(null, new Set(['kill_m_line10'])), null);
  assert.equal(salvageSource('', new Set(['kill_m_line10'])), null);
});

File: backend/src/test/java/tech/mikhailov/ijt/SalvageTest.java
package tech.mikhailov.ijt;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

/// A batch round writes N tests into one file. One bad assertion used to cost all N.
///
/// The very first unit of the v2 run proved it: three targets, two of them asserting wrongly,
/// the whole file deleted for breaking the suite, and the round recorded as cov 0→0 mut 0→0.
/// Under the old one-mutant-per-round loop a bad assertion cost one test; batching made it
/// cost the batch.
class SalvageTest {

    /// Braces inside string literals are not structure — the fixture below deliberately
    /// asserts on "}" — so balance is checked over CODE only. Counting raw characters made the
    /// UNMODIFIED fixture look unbalanced (4 vs 6) and failed a correct implementation. That
    /// was a bug in the test, not the code, and it is worth not repeating in the port.
    private static int[] braces(String s) {
        String c = JavaSrc.stripNonCode(s);
        return new int[]{
                (int) c.chars().filter(ch -> ch == '{').count(),
                (int) c.chars().filter(ch -> ch == '}').count()};
    }

    // captured verbatim from the live v2 run
    private static final String GRADLE = """
            ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest > kill_incrementBatchLoadExceptionCount_line76() FAILED
                at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
                at org.dataloader.stats.ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest.kill_incrementBatchLoadExceptionCount_line76(ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest.java:27)
            ThreadLocalStatisticsCollectorIncrementBatchLoadExceptionCountMacMutTest > kill_incrementBatchLoadExceptionCount_line82() FAILED
                at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)""";

    private static final String SUREFIRE = """
            [ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0
            [ERROR] kill_mustEscape_line170(org.json.XMLMacMutTest)  Time elapsed: 0.01 s  <<< FAILURE!
            java.lang.AssertionError: expected:<true> but was:<false>""";

    @Test
    void theFailingMethodNamesAreReadOutOfAGradleRun() {
        assertEquals(
                List.of("kill_incrementBatchLoadExceptionCount_line76", "kill_incrementBatchLoadExceptionCount_line82"),
                Salvage.failedTestNames(GRADLE).stream().sorted().toList());
    }

    @Test
    void andOutOfASurefireRun() {
        assertEquals(List.of("kill_mustEscape_line170"), List.copyOf(Salvage.failedTestNames(SUREFIRE)));
    }

    @Test
    void aGreenLogNamesNothing() {
        assertEquals(0, Salvage.failedTestNames("BUILD SUCCESSFUL in 12s").size());
        assertEquals(0, Salvage.failedTestNames("").size());
        assertEquals(0, Salvage.failedTestNames(null).size());
    }

    // ── cutting the bad methods out ───────────────────────────────────────
    private static final String FILE = """
            package a;

            import org.junit.jupiter.api.Test;
            import static org.hamcrest.MatcherAssert.assertThat;
            import static org.hamcrest.CoreMatchers.equalTo;

            class BMacMutTest {

                @Test
                void kill_m_line10() {
                    assertThat(new B().m(), equalTo(1));
                }

                @Test
                void kill_m_line20() {
                    // this one asserts the wrong thing
                    assertThat(new B().m(), equalTo(99));
                }

                @Test
                void kill_m_line30() {
                    assertThat(new B().n("}"), equalTo("}"));
                }
            }
            """;

    @Test
    void onlyTheNamedMethodIsRemoved() {
        String out = Salvage.dropTestMethods(FILE, Set.of("kill_m_line20"));
        assertFalse(out.contains("kill_m_line20"));
        assertTrue(out.contains("kill_m_line10"));
        assertTrue(out.contains("kill_m_line30"));
    }

    @Test
    void theFileStillHasItsPackageImportsAndClass() {
        String out = Salvage.dropTestMethods(FILE, Set.of("kill_m_line20"));
        assertTrue(out.startsWith("package a;"));
        assertTrue(out.contains("import org.junit.jupiter.api.Test;"));
        assertTrue(out.contains("class BMacMutTest {"));
        int[] b = braces(out);
        assertEquals(b[0], b[1], "braces balance");
    }

    @Test
    void aBraceInsideAStringLiteralDoesNotEndTheMethodEarly() {
        // kill_m_line30 asserts on "}" — counting braces naively would cut the file in half
        String out = Salvage.dropTestMethods(FILE, Set.of("kill_m_line30"));
        assertFalse(out.contains("kill_m_line30"));
        assertTrue(out.contains("kill_m_line10"));
        assertTrue(out.contains("kill_m_line20"));
        int[] b = braces(out);
        assertEquals(b[0], b[1], "braces balance");
    }

    @Test
    void removingEveryMethodLeavesACompilableEmptyClassNotABrokenFile() {
        String out = Salvage.dropTestMethods(FILE, Set.of("kill_m_line10", "kill_m_line20", "kill_m_line30"));
        assertTrue(out.contains("class BMacMutTest {"));
        assertFalse(out.contains("@Test"));
        int[] b = braces(out);
        assertEquals(b[0], b[1]);
    }

    @Test
    void namingAMethodThatIsNotThereChangesNothing() {
        assertEquals(FILE, Salvage.dropTestMethods(FILE, Set.of("kill_m_line999")));
        assertEquals(FILE, Salvage.dropTestMethods(FILE, Set.of()));
    }

    // ── one decision, used by both places that drop tests ─────────────────
    // The red-suite path learned to cut only the failing methods. The PIT green-suite
    // recovery did not, and deleted the whole file — so a batch that survived a red suite by
    // salvage was then wiped whole when PIT refused to run against one of its members. Two
    // callers, one decision, or they drift.
    @Test
    void cuttingTheNamedMethodsReturnsTheFileWhenSomethingStillRuns() {
        String out = Salvage.salvageSource(FILE, Set.of("kill_m_line20"));
        assertNotNull(out);
        assertFalse(out.contains("kill_m_line20"));
        assertTrue(out.contains("kill_m_line10"));
    }

    @Test
    void cuttingEverythingReturnsNullThereIsNothingWorthKeeping() {
        // a file whose only tests all failed is a file to delete, not to keep empty
        assertNull(Salvage.salvageSource(FILE, Set.of("kill_m_line10", "kill_m_line20", "kill_m_line30")));
    }

    @Test
    void namingNothingThatIsInTheFileReturnsNullNotANoOpKeep() {
        // the caller asked us to drop something we cannot find — deleting is the safe answer,
        // since the file demonstrably broke the build and we cannot say which part
        assertNull(Salvage.salvageSource(FILE, Set.of("kill_other_line1")));
        assertNull(Salvage.salvageSource(FILE, Set.of()));
    }

    @Test
    void aFileWeCannotReadIsNotSalvageable() {
        assertNull(Salvage.salvageSource(null, Set.of("kill_m_line10")));
        assertNull(Salvage.salvageSource("", Set.of("kill_m_line10")));
    }

    @Test
    void aMethodNameCarryingADollarIsMatchedLiterallyAndItsSiblingsSurvive() {
        // A DELIBERATE divergence from salvage.js, recorded here rather than reproduced.
        //
        // The JS interpolates the name straight into a regex — `new RegExp(`\\b${n}\\s*\\(`)` —
        // and `failedTestNames` deliberately admits `$` (its own pattern is `[A-Za-z_$][\w$]*`,
        // which is what a Java identifier allows). For a method called `a$b` the pattern becomes
        // `\ba$b\s*\(`, where `$` is an end-of-input anchor, so it can never match: `mine` comes
        // out empty, salvageSource returns null, and BOTH callers read that as "nothing worth
        // keeping" and delete the whole generated file — losing the sibling tests that passed.
        //
        // Pattern.quote here matches the name literally, so only the named method is cut. This
        // is not the JS behaviour, and it is not being made to be: the JS reading destroys work
        // the run has already paid for, and no caller wants it.
        String src = """
                class T {

                    @Test
                    void a$b() {
                        assertTrue(true);
                    }

                    @Test
                    void keep() {
                        assertTrue(true);
                    }
                }
                """;
        String out = Salvage.salvageSource(src, Set.of("a$b"));
        assertNotNull(out, "the JS answers null here and the whole file is deleted with it");
        assertFalse(out.contains("a$b"));
        assertTrue(out.contains("keep()"), "the sibling that passed is kept");
    }
}
